### PR TITLE
FLUME-2293. Isolate Flume agent plugins to their own classloader

### PR DESCRIPTION
**Pull request by ankitnanglia**

### CODE REVIEW

File: flume-ng-core/src/main/java/org/apache/flume/plugins/PluginsClassLoaderFactory.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      FileFilter jarFileFilter = new SuffixFileFilter(".jar");
+      ArrayList<URL> urls = new ArrayList<>();
+      for (File file : files) {
+        final File[] jarFiles = file.listFiles(jarFileFilter);
+        if (jarFiles != null) {
+          for (File jarFile : jarFiles) {
+            try {
+              urls.add(jarFile.toURI().toURL());
+            } catch (MalformedURLException e) {
+              logger.trace("Error adding jar file {} to class loader ",
```
Comment (Srinathc) [inline flume-ng-core/src/main/java/org/apache/flume/plugins/PluginsClassLoaderFactory.java:82]: log at ERROR level

**Review by Srinathc — COMMENTED** [review]
Does this include the backward compatibility changes?

File: flume-ng-core/src/main/java/org/apache/flume/plugins/PluginsClassLoaderFactory.java (hunk 1 above)
Comment (ankitnanglia) [reply flume-ng-core/src/main/java/org/apache/flume/plugins/PluginsClassLoaderFactory.java:82]: Forgot to push the change for backward compatibility, 
Backward compability is supported now

File: flume-ng-node/src/main/java/org/apache/flume/node/Application.java [hunk 2]
```diff
@@ -273,6 +279,12 @@ public static void main(String[] args) {
       if (commandLine.hasOption('z') || commandLine.hasOption("zkConnString")) {
         isZkConfigured = true;
       }
+
+      // initialize class loader factory from the plugins path
+      // this will isolate Flume agent plugins to their own classloader
+      String pluginsPath = commandLine.getOptionValue("plugins-path");
+      PluginsClassLoaderFactory.initialize(pluginsPath);
```
Comment (Srinathc) [inline flume-ng-node/src/main/java/org/apache/flume/node/Application.java:286]: shouldn't this be guarded by the --use-plugins-classloader flag?

Comment (ankitnanglia) [reply flume-ng-node/src/main/java/org/apache/flume/node/Application.java:286]: plugins-path arg is set only when --use-plugins-classloader is set to true. 
In normal flow, the plugins-path will be null resulting into all plugins being loaded by same Application class loader.

**Pull request closed without merging**